storage/testonly/fake_node_reader.go

```
@@ -0,0 +1,196 @@
+	return ok
+}
+
+// MultiFakeNodeReader can provide nodes at multiple revisions. It delegates to a number of
+// FakeNodeReaders, each set up to handle one revision.
+type MultiFakeNodeReader struct {
+	readers []FakeNodeReader
+}
+
+// LeafBatch describes a set of leaves to be loaded into a MultiFakeNodeReader via a compact
+// merkle tree. As each batch is added to the tree a set of node updates are collected
+// and recorded in a FakeNodeReader for that revision.
+type LeafBatch struct {
+	TreeRevision int64
+	Leaves       []string
+	ExpectedRoot string
+}
+
+// NewMultiFakeNodeReader creates a MultiFakeNodeReader delegating to a number of FakeNodeReaders
+func NewMultiFakeNodeReader(readers []FakeNodeReader) *MultiFakeNodeReader {
+	return &MultiFakeNodeReader{readers: readers}
+}
+
+// NewMultiFakeNodeReaderFromLeaves uses a compact Merkle tree to set up the nodes at various
+// revisions. It collates all node updates from a batch of leaf data into one FakeNodeReader.
+// This has the advantage of not needing to manually create all the data structures but the
+// disadvantage is that a bug in the compact tree could be reflected in test using this
+// code. To help guard against this we check the tree root hash after each batch has been
+// processed.
+func NewMultiFakeNodeReaderFromLeaves(batches []LeafBatch) *MultiFakeNodeReader {
+	tree := merkle.NewCompactMerkleTree(merkle.NewRFC6962TreeHasher(crypto.NewSHA256()))
+	readers := make([]FakeNodeReader, 0, len(batches))
+
+	for _, batch := range batches {
+		nodeMap := make(map[string]storage.Node)
+		for _, leaf := range batch.Leaves {
+			// We're only interested in the side effects of adding leaves - the node updates
+			tree.AddLeaf([]byte(leaf), func(depth int, index int64, hash []byte) {
+				nID, err := storage.NewNodeIDForTreeCoords(int64(depth), index, 64)
+
+				if err != nil {
+					panic(fmt.Errorf("failed to create a nodeID for tree - should not happen d:%d i:%d",
```

nit: In the grand scheme of things it doesn't matter much, but I would prefer a test failure to a panic. Maybe return an error, or pass in a *testing.T? (Not sure if the latter is idiomatic either.)

Feel free to push back.